src/main/java/org/team1294/firstpowerup/robot/subsystems/VisionSubsystem.java

```diff
@@ -0,0 +1,191 @@
+import edu.wpi.cscore.UsbCamera;
+import edu.wpi.first.wpilibj.CameraServer;
+import edu.wpi.first.wpilibj.command.Subsystem;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import org.opencv.core.Mat;
+import org.opencv.core.Rect;
+import org.opencv.core.Scalar;
+import org.opencv.imgproc.Imgproc;
+import org.team1294.firstpowerup.robot.Robot;
+import org.team1294.firstpowerup.robot.commands.DefaultVisionCommand;
+import org.team1294.firstpowerup.robot.vision.CratePipeline;
+import org.team1294.firstpowerup.robot.vision.PairOfRect;
+import org.team1294.firstpowerup.robot.vision.SwitchTargetPipeline;
+import org.team1294.firstpowerup.robot.vision.VisionProcessingResult;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class VisionSubsystem extends Subsystem {
+
+    private static final int IMG_WIDTH = 320;
+    private static final int IMG_HEIGHT = 240;
+    private final CvSink cvSink;
+    private final CvSource cvSource;
+    private final UsbCamera usbCamera;
+    private final CameraServer cameraServer;
+    private final Mat frame = new Mat();
+    private final SwitchTargetPipeline switchTargetPipeline;
+    private final CratePipeline cratePipeline;
+
+    private VisionProcessingResult visionProcessingResult;
+
+    public VisionSubsystem() {
+        switchTargetPipeline = new SwitchTargetPipeline();
+        cratePipeline = new CratePipeline();
+
+        cameraServer = CameraServer.getInstance();
+
+        usbCamera = cameraServer.startAutomaticCapture(0);
+        usbCamera.setResolution(IMG_WIDTH, IMG_HEIGHT);
+        usbCamera.setFPS(30);
+
+        cvSink = cameraServer.getVideo(usbCamera);
+        cvSource = cameraServer.putVideo("VisionSystem", IMG_WIDTH, IMG_HEIGHT);
+
+        //https://wpilib.screenstepslive
+        // .com/s/currentCS/m/vision/l/669166-using-the-cameraserver-on-the-roborio
+        new Thread(() -> {
+            // grab a frame from the camera
+            cvSink.grabFrame(frame);
+
+            // draw on the frame if we have a vision target
+            if (visionProcessingResult != null && visionProcessingResult.isTargetAcquired()) {
+                final Rect targetRect = visionProcessingResult.getTargetRect();
+                Imgproc
+                    .rectangle(frame, targetRect.tl(), targetRect.br(), new Scalar(0, 0, 255), 2);
+            }
+
+            // output the frame to the stream
+            cvSource.putFrame(frame);
+        }).start();
+    }
+
+
+    @Override
+    protected void initDefaultCommand() {
+        setDefaultCommand(new DefaultVisionCommand());
+    }
+
+    public VisionProcessingResult detectNothing() {
+        // set the camera to take normal images
+        setCameraNormal();
+
+        // create and return a negative result
+        final VisionProcessingResult result = new VisionProcessingResult();
+        this.visionProcessingResult = result;
```

Comment:
Why does the result need to be saved as a field here? Anything that needs to manipulate it should do so before it's returned, and a client of these methods should be saving the result themselves.

Reply:
because of the thread created in the constructor that draws a rectangle around the successful vision processing result.